File: sdk/typescript/src/__tests__/mail.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import { MashgateClient } from "../client.js";
import { MessageFolder, MailboxStatus } from "../resources/mail.js";

interface MockCall {
  url: string;
  init: RequestInit & { headers: Record<string, string> };
}

function mockFetchReturning(body: unknown, status = 200) {
  return vi.fn().mockResolvedValue({
    ok: status >= 200 && status < 300,
    status,
    statusText: "OK",
    headers: new Headers(),
    json: () => Promise.resolve(body),
  });
}

function lastCall(mock: ReturnType<typeof mockFetchReturning>): MockCall {
  const [url, init] = mock.mock.calls[mock.mock.calls.length - 1];
  return { url: String(url), init: init as MockCall["init"] };
}

describe("MailResource", () => {
  let mockFetch: ReturnType<typeof mockFetchReturning>;
  let client: MashgateClient;

  beforeEach(() => {
    mockFetch = mockFetchReturning({});
    client = new MashgateClient({
      baseUrl: "https://api.mashgate.uz",
      apiKey: "mg_test_key",
      fetch: mockFetch,
    });
  });

  it("getMyMailbox GETs /v1/mail/mailboxes/me", async () => {
    mockFetch = mockFetchReturning({
      mailbox_id: "mb-1",
      tenant_id: "tnt-1",
      subject_id: "u-1",
      email: "demo@mail.entry-i.com",
      status: "MAILBOX_STATUS_ACTIVE",
      quota_bytes: 5368709120,
      used_bytes: 0,
      created_at: "2026-04-27T00:00:00Z",
      updated_at: "2026-04-27T00:00:00Z",
    });
    client = new MashgateClient({
      baseUrl: "https://api.mashgate.uz",
      apiKey: "mg_test_key",
      fetch: mockFetch,
    });

    const mb = await client.mail.getMyMailbox();
    expect(mb.email).toBe("demo@mail.entry-i.com");
    expect(mb.status).toBe(MailboxStatus.Active);

    const { url, init } = lastCall(mockFetch);
    expect(url).toBe("https://api.mashgate.uz/v1/mail/mailboxes/me");
    expect(init.method).toBe("GET");
  });

  it("listMessages passes folder and pagination as query", async () => {
    await client.mail.listMessages({
      folder: MessageFolder.Sent,
      limit: 25,
      cursor: "abc",
    });
    const { url } = lastCall(mockFetch);
    expect(url).toContain("/v1/mail/messages");
    expect(url).toContain("folder=MESSAGE_FOLDER_SENT");
    expect(url).toContain("limit=25");
    expect(url).toContain("cursor=abc");
  });

  it("sendMessage POSTs body to /v1/mail/messages with idempotency_key", async () => {
    mockFetch = mockFetchReturning({
      message_id: "msg-1",
      status: "SEND_STATUS_QUEUED",
      queued_at: "2026-04-27T10:00:00Z",
    });
    client = new MashgateClient({
      baseUrl: "https://api.mashgate.uz",
      apiKey: "mg_test_key",
      fetch: mockFetch,
    });

    const out = await client.mail.sendMessage({
      to: ["alice@example.com"],
      subject: "hi",
      body_text: "test",
      idempotency_key: "send-1",
    });
    expect(out.message_id).toBe("msg-1");
    expect(out.status).toBe("SEND_STATUS_QUEUED");

    const { url, init } = lastCall(mockFetch);
    expect(url).toBe("https://api.mashgate.uz/v1/mail/messages");
    expect(init.method).toBe("POST");
    const body = JSON.parse(init.body as string);
    expect(body.to).toEqual(["alice@example.com"]);
    expect(body.subject).toBe("hi");
    expect(body.idempotency_key).toBe("send-1");
  });

  it("updateMessage PATCHes only provided fields", async () => {
    await client.mail.updateMessage("msg-1", {
      read: true,
      folder: MessageFolder.Trash,
    });
    const { url, init } = lastCall(mockFetch);
    expect(url).toBe("https://api.mashgate.uz/v1/mail/messages/msg-1");
    expect(init.method).toBe("PATCH");
    const body = JSON.parse(init.body as string);
    expect(body.read).toBe(true);
    expect(body.folder).toBe(MessageFolder.Trash);
    expect(body.labels).toBeUndefined();
  });

  it("deleteMessage soft-deletes by default", async () => {
    await client.mail.deleteMessage("msg-1");
    const { url, init } = lastCall(mockFetch);
    expect(url).toContain("/v1/mail/messages/msg-1");
    expect(url).toContain("hard_delete=false");
    expect(init.method).toBe("DELETE");
  });

  it("createDomain POSTs name and returns Domain in pending", async () => {
    mockFetch = mockFetchReturning({
      domain_id: "d-1",
      tenant_id: "tnt-1",
      name: "mail.entry-i.com",
      status: "DOMAIN_STATUS_PENDING",
      mx_records: [],
      created_at: "2026-04-27T00:00:00Z",
    });
    client = new MashgateClient({
      baseUrl: "https://api.mashgate.uz",
      apiKey: "mg_test_key",
      fetch: mockFetch,
    });

    const d = await client.mail.createDomain({ name: "mail.entry-i.com" });
    expect(d.status).toBe("DOMAIN_STATUS_PENDING");

    const { url, init } = lastCall(mockFetch);
    expect(url).toBe("https://api.mashgate.uz/v1/mail/domains");
    expect(init.method).toBe("POST");
    const body = JSON.parse(init.body as string);
    expect(body.name).toBe("mail.entry-i.com");
  });

  it("rotateDKIM defaults to 2048 bits", async () => {
    await client.mail.rotateDKIM("d-1");
    const { url, init } = lastCall(mockFetch);
    expect(url).toBe("https://api.mashgate.uz/v1/mail/domains/d-1/dkim/rotate");
    expect(init.method).toBe("POST");
    const body = JSON.parse(init.body as string);
    expect(body.key_bits).toBe(2048);
  });
});

File: sdk/go/mail_test.go
package mashgate

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"
	"testing"
	"time"
)

func TestMail_GetMyMailbox(t *testing.T) {
	want := Mailbox{
		MailboxID:   "mb-1",
		TenantID:    "tnt-1",
		SubjectID:   "u-1",
		Email:       "demo@mail.entry-i.com",
		Status:      MailboxStatusActive,
		QuotaBytes:  5_368_709_120,
		UsedBytes:   42_342_342,
		DisplayName: "Demo User",
		CreatedAt:   time.Date(2026, 4, 27, 0, 0, 0, 0, time.UTC),
		UpdatedAt:   time.Date(2026, 4, 27, 0, 0, 0, 0, time.UTC),
	}

	_, client := newTestServerFn(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/mail/mailboxes/me" {
			t.Errorf("expected path /v1/mail/mailboxes/me, got %s", r.URL.Path)
		}
		if r.Method != http.MethodGet {
			t.Errorf("expected GET, got %s", r.Method)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(want)
	})

	got, err := client.Mail.GetMyMailbox(context.Background())
	if err != nil {
		t.Fatalf("GetMyMailbox: %v", err)
	}
	if got.Email != want.Email || got.Status != want.Status {
		t.Errorf("mismatch: got %+v, want %+v", got, want)
	}
}

func TestMail_ListMessages_query(t *testing.T) {
	_, client := newTestServerFn(t, func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		if q.Get("folder") != string(MessageFolderSent) {
			t.Errorf("expected folder=SENT, got %q", q.Get("folder"))
		}
		if q.Get("limit") != "25" {
			t.Errorf("expected limit=25, got %q", q.Get("limit"))
		}
		if q.Get("cursor") != "abc" {
			t.Errorf("expected cursor=abc, got %q", q.Get("cursor"))
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(ListMailMessagesResponse{Items: []MailMessagePreview{}})
	})

	_, err := client.Mail.ListMessages(context.Background(), ListMailMessagesQuery{
		Folder: MessageFolderSent,
		Limit:  25,
		Cursor: "abc",
	})
	if err != nil {
		t.Fatalf("ListMessages: %v", err)
	}
}

func TestMail_SendMessage_idempotency(t *testing.T) {
	_, client := newTestServerFn(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/mail/messages" || r.Method != http.MethodPost {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		var body SendMailRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("decode body: %v", err)
		}
		if body.IdempotencyKey != "send-1" {
			t.Errorf("expected idempotency_key send-1, got %q", body.IdempotencyKey)
		}
		if len(body.To) != 1 || body.To[0] != "alice@example.com" {
			t.Errorf("unexpected to=%v", body.To)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(SendMailResponse{
			MessageID: "msg-1",
			Status:    SendStatusQueued,
			QueuedAt:  time.Now(),
		})
	})

	out, err := client.Mail.SendMessage(context.Background(), SendMailRequest{
		To:             []string{"alice@example.com"},
		Subject:        "hi",
		BodyText:       "test",
		IdempotencyKey: "send-1",
	})
	if err != nil {
		t.Fatalf("SendMessage: %v", err)
	}
	if out.Status != SendStatusQueued || out.MessageID != "msg-1" {
		t.Errorf("unexpected response: %+v", out)
	}
}

func TestMail_DeleteMessage_softByDefault(t *testing.T) {
	_, client := newTestServerFn(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete {
			t.Errorf("expected DELETE, got %s", r.Method)
		}
		if !strings.Contains(r.URL.RawQuery, "hard_delete=false") {
			t.Errorf("expected hard_delete=false, got %s", r.URL.RawQuery)
		}
		w.WriteHeader(http.StatusNoContent)
	})
	if err := client.Mail.DeleteMessage(context.Background(), "msg-1", false); err != nil {
		t.Fatalf("DeleteMessage: %v", err)
	}
}

func TestMail_RotateDKIM_defaultsTo2048(t *testing.T) {
	_, client := newTestServerFn(t, func(w http.ResponseWriter, r *http.Request) {
		var body RotateDKIMRequest
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body.KeyBits != 2048 {
			t.Errorf("expected default key_bits=2048, got %d", body.KeyBits)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(MailDomain{
			DomainID: "d-1",
			Status:   DomainStatusActive,
		})
	})
	_, err := client.Mail.RotateDKIM(context.Background(), "d-1", 0)
	if err != nil {
		t.Fatalf("RotateDKIM: %v", err)
	}
}
